crates/wasmparser/src/readers/core/types.rs

```diff
@@ -522,53 +509,42 @@ impl RefType {
 
     // Note that this is similar to `Display for RefType` except that it has
     // the indexes stubbed out.
-    pub(crate) fn wat(&self) -> &'static str {
+    pub(crate) fn wat(&self) -> Cow<'static, str> {
         match (self.is_nullable(), self.heap_type()) {
-            (true, HeapType::Func) => "funcref",
-            (true, HeapType::Extern) => "externref",
-            (true, HeapType::Indexed(_)) => "(ref null $type)",
-            (true, HeapType::Any) => "anyref",
-            (true, HeapType::None) => "nullref",
-            (true, HeapType::NoExtern) => "nullexternref",
-            (true, HeapType::NoFunc) => "nullfuncref",
-            (true, HeapType::Eq) => "eqref",
-            (true, HeapType::Struct) => "structref",
-            (true, HeapType::Array) => "arrayref",
-            (true, HeapType::I31) => "i31ref",
-            (false, HeapType::Func) => "(ref func)",
-            (false, HeapType::Extern) => "(ref extern)",
-            (false, HeapType::Indexed(_)) => "(ref $type)",
-            (false, HeapType::Any) => "(ref any)",
-            (false, HeapType::None) => "(ref none)",
-            (false, HeapType::NoExtern) => "(ref noextern)",
-            (false, HeapType::NoFunc) => "(ref nofunc)",
-            (false, HeapType::Eq) => "(ref eq)",
-            (false, HeapType::Struct) => "(ref struct)",
-            (false, HeapType::Array) => "(ref array)",
-            (false, HeapType::I31) => "(ref i31)",
+            (true, HeapType::Func) => "funcref".into(),
+            (true, HeapType::Extern) => "externref".into(),
+            (true, HeapType::Indexed(i)) => format!("(ref null {i})").into(),
```

Comment:
IIRC this `wat()` representation was used in some perf-sensitive situations in validation, namely around it created bloat around error messages I think. I forget the exact context but that's why it was the way it was before. If you're ok leaving this as it was before I think that'd be best, otherwise though I can go look into what it was and see if it's still important